src/osd/scrub_machine.cc

```diff
@@ -0,0 +1,768 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#include "scrub_machine.h"
+
+#include <chrono>
+
+#include "OSD.h"
+#include "OpRequest.h"
+#include "ScrubStore.h"
+#include "pg_scrubber.h"  // RRR try to remove this one
+#include "scrub_machine_lstnr.h"
+
+#define dout_context g_ceph_context
+#define dout_subsys ceph_subsys_osd
+#undef dout_prefix
+#define dout_prefix *_dout << " scrbrFSM.(~~)"
+
+
+using namespace std::chrono;
+using namespace std::chrono_literals;
+namespace sc = boost::statechart;
+
+#define DECLARE_LOCALS                                          \
+  ScrubMachineListener* scrbr = context<ScrubMachine>().scrbr_; \
+  std::ignore = scrbr;                                          \
+  PG* pg = context<ScrubMachine>().pg_;                         \
+  std::ignore = pg;                                             \
+  auto pg_id = context<ScrubMachine>().pg_id_;                  \
+  std::ignore = pg_id;
+
+namespace Scrub {
+
+//
+//  trace/debug auxiliaries
+//
+
+state_logger_t::state_logger_t(std::string nm) : name_{nm}
+{
+  dout(7) << " ------ state -->> " << name_ << dendl;
+}
+
+void on_event_creation(std::string nm)
```

Comment:
const &?

Reply:
that''s dev code, not even testing. Will be removed.

Reply:
made it string_view